lib/oper.g

```diff
@@ -615,6 +615,85 @@ BIND_GLOBAL( "TraceImmediateMethods", function( arg )
     fi;
 end );
 
+#############################################################################
+##
+##
+##  <#GAPDoc Label="TraceInternalMethods">
+##  <ManSection>
+##  <Func Name="TraceInternalMethods" Arg=''/>
+##  <Func Name="UntraceInternalMethods" Arg=''/>
+##  <Func Name="GetTracedInternalMethodsCounts" Arg=''/>
+##  <Func Name="ClearTraceInternalMethodsCounts" Arg=''/>
+##
+##  <Description>
+##  <Ref Func="TraceInternalMethods"/> enables tracing for all internal methods.
+##  Internal methods are methods which implement many fundamental operations in GAP.
+##  In this version of GAP, the internal methods which can be traced is:
+##  <List>
+##  <Mark>Zero, ZeroMut</Mark><Item>Mutable and Immutable <Ref Attr="Zero"/></Item>
+##  <Mark>AInv, AInvMut</Mark><Item>Mutable and Immutable <Ref Attr="AdditiveInverse"/></Item>
+##  <Mark>One, OneMut</Mark><Item>Mutable and Immutable <Ref Attr="One"/></Item>
+##  <Mark>Inv, InvMut</Mark><Item>Mutable and Immutable <Ref Attr="Inverse"/></Item>
+##  <Mark>Sum</Mark><Item>The operator <Ref Oper="\+"/></Item>
+##  <Mark>Diff</Mark><Item>The operator <C>-</C> operator</Item>
+##  <Mark>Prod</Mark><Item>The operator <Ref Oper="\*"/></Item>
+##  <Mark>Quo</Mark><Item>The operator <Ref Oper="\/"/></Item>
+##  <Mark>LQuo</Mark><Item>The left-quotient operator</Item>
+##  <Mark>Pow</Mark><Item>The operator <Ref Oper="\^"/></Item>
+##  <Mark>Comm</Mark><Item>The operator <Ref Oper="Comm"/></Item>
+##  <Mark>Mod</Mark><Item>The operator <Ref Oper="\mod"/></Item>
+##  </List>
+##  <P/>
+##  <Ref Func="UntraceInternalMethods"/> turns tracing off.
+##  As these methods can be called hundreds of thousands of times in simple GAP
+##  code, there isn't a statement printed each time one is called. Instead, the
+##  method <Ref Func="GetTracedInternalMethodsCounts"/> returns how many times
+##  each operation has been applied to each type of variable (the type of a
+##  variable can be found with the <C>TNUM_OBJ</C> method).
+##  <Log><![CDATA[
+## gap> TraceInternalMethods();
+## true
+## gap> 2+3+4+5+6;;
+## gap> 2.0+2.0;
+## gap> GetTracedInternalMethodsCounts();
+## rec( Sum := rec( 0 := rec( 0 := 4 ), 6 := rec( 6 := 1 ) ) )
```

Comment:
To be honest I am not sure if it is clear how to interpret this output. So, in your example
`GetTracedInternalMethodsCounts().sum;` yields `rec( 0 := rec( 0 := 4 ), 6 := rec( 6 := 1 ) )`. Why are the entries of this themselves records again? The first entry concerns integers (TNUM type 0), and it is a record with entry `0:=4`. Maybe this could be made clearer in the documentation? If I'm the only one that is not happy with this output I'm happy to remove this comment though.

Reply:
I can certainly explain better. There are two parts:

Sum is a 2-argument operator, so Sum.0.0 tells you how many times sum got called with 2 arguments, the first of TNUM 0 and the second of TNUM 0.

Why records instead of lists? Because a list of lists here can end up being really big, with mostly unbound values (or full of zeroes, depending on how you treat operators which never get called), which is annoying to read. However, it might still be better to just use a list of lists, as that is a more natural structure.